dafni_cli/datasets/dataset_upload.py

```diff
@@ -15,37 +16,66 @@
 )
 from dafni_cli.api.session import DAFNISession
 
+# Keys inside dataset metadata returned from the API that are invalid for
+# uploading
+METADATA_KEYS_INVALID_FOR_UPLOAD = [
+    "@id",
+    "dct:issued",
+    "dct:modified",
+    "mediatypes",
+    "version_history",
+    "auth",
+    "dcat:distribution",
+]
 
-def upload_new_dataset_files(
-    session: DAFNISession, definition_path: Path, file_paths: List[Path]
-) -> None:
-    """Function to upload all files associated with a new Dataset
+
+def remove_dataset_metadata_invalid_for_upload(metadata: dict):
+    """Function to remove metadata for a dataset that is given when getting it
+    but are not valid during upload
 
     Args:
-        session (DAFNISession): User session
-        definition_path (Path): Path to Dataset metadata file
-        file_paths (List[Path]): List of Paths to dataset data files
+        metadata (dict): The metadata to modify
     """
 
-    click.echo("\nRetrieving temporary bucket ID")
-    temp_bucket_id = create_temp_bucket(session)
+    for key in METADATA_KEYS_INVALID_FOR_UPLOAD:
+        del metadata[key]
 
-    # If any exception happens now, we want to make sure we delete the
-    # temporary bucket to prevent a build up in the user's quota
-    try:
-        # Upload all files
-        upload_files(session, temp_bucket_id, file_paths)
-        details = upload_metadata(session, definition_path, temp_bucket_id)
-    except BaseException:
-        click.echo("Deleting temporary bucket")
-        delete_temp_bucket(session, temp_bucket_id)
-        raise
 
-    # Output Details
-    click.echo("\nUpload successful")
-    click.echo(f"Dataset ID: {details['datasetId']}")
-    click.echo(f"Version ID: {details['versionId']}")
-    click.echo(f"Metadata ID: {details['metadataId']}")
+def modify_dataset_metadata_for_upload(
+    existing_metadata: dict,
+    definition_path: Optional[Path],
+    version_message: Optional[str],
+) -> dict:
+    """Modifies existing dataset metadata or that loaded from a file according
+    to user specified parameters and in a format ready for upload
+
+    Args:
+        existing_metadata (dict): Dictionary of existing metadata from the API
+        definition_path (Optional[Path]): Path to a Dataset metadata file.
+                        When None will use the existing_metadata instead but
+                        will delete any keys invalid for upload.
+        version_message (Optional[str]): Version message - Will replace
+                        whatever already exists in the loaded metadata
+    Returns:
+        dict: The modified dataset metadata
+    """
+
+    # Load the metadata from the definition file if present, or otherwise
```

Comment:
It may be possible that a user uses the save function to get the current metadata, makes changes in the file but doesn't remove the unwanted fields, then tries to re-upload - it may be worth running remove_dataset_metadata_invalid_for_upload regardless of how the metadata is loaded